src/subscription/subscribe.js

```diff
@@ -0,0 +1,161 @@
+import mapAsyncIterator from './mapAsyncIterator';
+
+import type {
+  ExecutionContext,
+  ExecutionResult,
+} from '../execution/execute';
+import type {
+  DocumentNode,
+  OperationDefinitionNode,
+} from '../language/ast';
+
+export function getSubscriptionEventSource(
+  schema: GraphQLSchema,
+  document: DocumentNode,
+  rootValue?: mixed,
+  contextValue?: mixed,
+  variableValues?: ?{[key: string]: mixed},
+  operationName?: ?string,
+): AsyncIterable<mixed> {
+  // If a valid context cannot be created due to incorrect arguments,
+  // this will throw an error.
+  const exeContext = buildExecutionContext(
+    schema,
+    document,
+    rootValue,
+    contextValue,
+    variableValues,
+    operationName
+  );
+
+  // Call the `subscribe()` resolver or the default resolver to produce an
+  // AsyncIterable yielding raw payloads.
+  return resolveSubscription(
+    exeContext,
+    exeContext.operation,
+    rootValue
+  );
+}
+
+/**
+ * Implements the "Subscribing to request" section of the GraphQL specification.
+ *
+ * Returns an AsyncIterator
+ *
+ * If the arguments to this function do not result in a legal execution context,
+ * a GraphQLError will be thrown immediately explaining the invalid input.
+ */
+export function subscribe(
+  schema: GraphQLSchema,
+  document: DocumentNode,
+  rootValue?: mixed,
+  contextValue?: mixed,
+  variableValues?: ?{[key: string]: mixed},
+  operationName?: ?string,
+): AsyncIterator<ExecutionResult> {
+  const subscription = getSubscriptionEventSource(
+    schema,
+    document,
+    rootValue,
+    contextValue,
+    variableValues,
+    operationName);
+
+  // For each payload yielded from a subscription, map it over the normal
+  // GraphQL `execute` function, with `payload` as the rootValue.
+  return mapAsyncIterator(
+    subscription,
+    payload => execute(
+      schema,
+      document,
+      payload,
+      contextValue,
+      variableValues,
+      operationName
+    )
+  );
+}
+
+function resolveSubscription(
+  exeContext: ExecutionContext,
+  operation: OperationDefinitionNode,
+  rootValue: mixed
+): AsyncIterable<mixed> {
+  const type = getOperationRootType(exeContext.schema, exeContext.operation);
+  const fields = collectFields(
+    exeContext,
+    type,
+    exeContext.operation.selectionSet,
+    Object.create(null),
+    Object.create(null)
+  );
+  const responseNames = Object.keys(fields);
+  invariant(
+    responseNames.length === 1,
+    'A subscription must contain exactly one field.'
+  );
+  const responseName = responseNames[0];
+  const fieldNodes = fields[responseName];
+  const fieldNode = fieldNodes[0];
+  const fieldDef = getFieldDef(exeContext.schema, type, fieldNode.name.value);
+  invariant(
+    fieldDef,
+    'This subscription is not defined by the schema.'
+  );
+
+  // TODO: make GraphQLSubscription flow type special to support defining these?
+  const resolveFn = (fieldDef: any).subscribe || defaultFieldResolver;
```

Comment:
I've added `subscribe` to the Flow types. 
Are you sure that `isValidResolver` can help us here? because `isValidResolver` allows `null` or `function` as resolvers, but in this case we want to enforce a resolver (and disallow `null` value).
In our case we want to make sure that `subscribe` resolver exists and returns a valid `AsyncIterable`, otherwise we can't resolve the subscription.